bindings/pubsubadapter/pubsubadapter.go

```diff
@@ -0,0 +1,150 @@
+	defaultInvokeTopicAttribute = "publishTopic"
+	defaultInvokeTopicMetadata  = "topic"
+)
+
+// Ensure Adapter satisfies the input and output binding interfaces.
+var (
+	_ bindings.InputBinding  = (*Adapter)(nil)
+	_ bindings.OutputBinding = (*Adapter)(nil)
+)
+
+// Adapter allows any PubSub component to be used as an input
+// and output binding.
+type Adapter struct {
+	name                 string
+	broker               pubsub.PubSub
+	readTopicAttribute   string
+	readTopic            string
+	invokeTopicAttribute string
+	invokeTopicMetadata  string
+	invokeTopic          string
+}
+
+// Option is passed into `New` and overrides a default value.
+type Option func(a *Adapter)
+
+// WithReadTopicAttribute overrides the default read topic attribute
+// that determines the topic to subscribe to when Read is called.
+func WithReadTopicAttribute(readTopicAttribute string) Option {
+	return func(a *Adapter) {
+		a.readTopicAttribute = readTopicAttribute
+	}
+}
+
+// WithInvokeTopicAttribute overrides the default invoke topic attribute
+// that determines the topic to publish to when Invoke is called and
+// the topic is not overridden in the request metadata.
+func WithInvokeTopicAttribute(invokeTopicAttribute string) Option {
+	return func(a *Adapter) {
+		a.invokeTopicAttribute = invokeTopicAttribute
+	}
+}
+
+// WithInvokeTopicMetadata overrides the default invoke topic metadata
+// name that determines the metadata property that overrides the topic
+// to publish to when Invoke is called.
+func WithInvokeTopicMetadata(invokeTopicMetadata string) Option {
+	return func(a *Adapter) {
+		a.invokeTopicMetadata = invokeTopicMetadata
+	}
+}
+
+// New returns an input/output binding adapter for `broker`.
+func New(broker pubsub.PubSub, options ...Option) *Adapter {
+	a := Adapter{ //nolint:exhaustivestruct
+		broker:               broker,
+		readTopicAttribute:   defaultReadTopicAttribute,
+		invokeTopicAttribute: defaultInvokeTopicAttribute,
+		invokeTopicMetadata:  defaultInvokeTopicMetadata,
+	}
+
+	for _, opt := range options {
+		opt(&a)
+	}
+
+	return &a
+}
+
+// Init reads the topic property and passes the properties
+// to the underlying PubSub component.
+func (a *Adapter) Init(metadata bindings.Metadata) error {
+	if t, ok := metadata.Properties[a.readTopicAttribute]; ok && t != "" {
+		a.readTopic = t
+	} else {
+		return fmt.Errorf("pubsub adapter: %q attribute was missing", a.readTopicAttribute)
+	}
+
+	if t, ok := metadata.Properties[a.invokeTopicAttribute]; ok && t != "" {
+		a.invokeTopic = t
+	} else {
+		return fmt.Errorf("pubsub adapter: %q attribute was missing", a.invokeTopicAttribute)
+	}
+
+	a.name = metadata.Name
+
+	return a.broker.Init(pubsub.Metadata{ //nolint:wrapcheck
+		Properties: metadata.Properties,
+	})
+}
+
+// Read subscribes to the configured topic and invokes `handler`
+// when a message is received.
+func (a *Adapter) Read(handler func(*bindings.ReadResponse) ([]byte, error)) error {
```

Comment:
Bindings have `Read()` as a blocking function where `Subscribe()` is not. That probably needs to be preserved.

Reply:
Great point

Reply:
I think this is OK because the runtime starts a goroutine to handle calling `Read()`. In the adapter scenario, `Read()` would return immediately (no error) but the Subscribe goroutine would be running and log any errors. Same net affect as far as I can tell.

https://github.com/dapr/dapr/blob/42a901540f0f5a28a618721471ad208dfbf04a19/pkg/runtime/runtime.go#L2217-L2227